quantity/core.py

```diff
@@ -0,0 +1,243 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import annotations
+
+import operator
+from dataclasses import dataclass, replace
+from typing import TYPE_CHECKING
+
+import array_api_compat
+import astropy.units as u
+import numpy as np
+from astropy.units.quantity_helper.converters import UFUNC_HELPERS
+
+if TYPE_CHECKING:
+    from typing import Any
+
+
+DIMENSIONLESS = u.dimensionless_unscaled
+
+PYTHON_NUMBER = float | int | complex
+
+
+def has_array_namespace(arg):
+    try:
+        array_api_compat.array_namespace(arg)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+def get_value_and_unit(arg, default_unit=None):
+    # HACK: interoperability with astropy Quantity.  Have protocol?
+    try:
+        unit = arg.unit
+    except AttributeError:
+        return arg, default_unit
+    else:
+        return arg.value, unit
+
+
+@dataclass(frozen=True, eq=False)
+class Quantity:
+    value: Any
+    unit: u.UnitBase
+
+    def _to_own_unit(self, value):
```

Comment:
Does this need to be a method so that subclasses may override it? If not, then this feels like a separate utility function. e.g.

```python
def as_quantity(value, unit):
    value, from_unit = get_value_and_unit(value, default_unit=DIMENSIONLESS)
    return from_unit.to(unit, value)
```

Reply:
Good point, let me take it out indeed - the fewer methods we put on `Quantity`, the better (and this in particular may be something more for the unit system).